File: go/206-ReverseLinkedList/main.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseList(head *ListNode) *ListNode {
	if head == nil {
		return head
	}

	var newHead *ListNode = nil

	for head != nil {
		after := head.Next
		head.Next = newHead
		newHead = head
		head = after
	}

	return newHead
}

func main() {
	list := &ListNode{Val: 1, Next: &ListNode{Val: 2, Next: &ListNode{Val: 3}}}
	reverseList(list)
}

File: cpp/206-ReverseLinkedList/main.cpp
struct ListNode 
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x, ListNode *next = nullptr) : val(x), next(next) {}
};

class Solution 
{
public:
    ListNode* reverseList(ListNode* head) 
    {
        ListNode *newHead = nullptr;

        while(head)
        {
            auto after = head->next;
            head->next = newHead;
            newHead = head;
            head = after;
        }

        return newHead;
    }
};

int main()
{
    Solution s;
    auto test1 = new ListNode(1, new ListNode(2, new ListNode(3)));
    s.reverseList(test1);
}